presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -61,13 +62,17 @@
 
     private final SqlStageExecution stage;
     private final List<InternalNode> nodes;
-    private final List<SourceScheduler> sourceSchedulers;
+
+    // use a CopyOnWriteArrayList to prevent ConcurrentModificationExceptions
+    // if close() is called while the main thread is in the scheduling loop
+    private final CopyOnWriteArrayList<SourceScheduler> sourceSchedulers;
     private final List<ConnectorPartitionHandle> partitionHandles;
     private boolean scheduledTasks;
     private boolean anySourceSchedulingFinished;
     private final Optional<LifespanScheduler> groupedLifespanScheduler;
 
     private final Queue<Integer> tasksToRecover = new ConcurrentLinkedQueue<>();
+    private volatile boolean cancelled;
```

Comment:
Should this be called `closed`?

Reply:
I wonder if we can simplify everything here by just making `cancelled` (`closed`) an `AtomicBoolean`, removing the `clear()` method invocation, and removing the `synchronized` blocks.

Reply:
Good question.  I'm actually not sure why it was being cleared in the first place.  My only thought is so that the schedulers get garbage collected.  Do you think if we don't clear the list the objects would hang around too long?

Also, we do need to exit the scheduler loop after closing because otherwise we hit "HiveSplitSource is already closed" errors.

Reply:
It could be for GC purposes, though the git blame doesn't suggest it.  ¯\_(ツ)_/¯  I don't think my comment is valid though, because I overlooked the .remove(), though I have a separate observation about that here: https://github.com/prestodb/presto/pull/13717#discussion_r373159482

Reply:
I got confused by the indentation in GH, makes sense.  I thought the synchronized block didn't include the remove.